File: CS/ZipReportStorage.cs
using System;
using System.IO;
using System.Windows.Forms;
using System.ComponentModel;
using System.IO.Compression;
using System.Collections.Generic;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;        
using DevExpress.XtraReports.UI;                 
using DevExpress.XtraReports.Extensions;           
// ...

namespace ReportStorageSample {
    class ZipReportStorage : ReportStorageExtension {
        const string fileName = "ReportStorage.zip";
        public ZipReportStorage() {
        }
        string StoragePath {
            get {
                string dirName = Path.GetDirectoryName(Application.ExecutablePath);
                return Path.Combine(dirName, fileName);
            }
        }
        public override bool CanSetData(string url) {
            // Always return true to confirm that the SetData method is available.
            return true;
        }
        public override bool IsValidUrl(string url) {
            return !string.IsNullOrEmpty(url);
        }
        public override byte[] GetData(string url) {
            // Open ZIP archive.
            if (!File.Exists(StoragePath))
                return new byte[] { };
            using (var fileStream = File.OpenRead(StoragePath))
            using (var archive = new ZipArchive(fileStream, ZipArchiveMode.Read)) {
                var entry = archive.GetEntry(url);
                if (entry == null)
                    return new byte[] { };
                using (var entryStream = entry.Open())
                using (var ms = new MemoryStream()) {
                    entryStream.CopyTo(ms);
                    return ms.ToArray();
                }
            }
        }
        static bool StringsEgual(string a, string b) {
            return string.Equals(a, b, StringComparison.OrdinalIgnoreCase);
        }
        public override void SetData(XtraReport report, string url) {
            report.Extensions["StorageID"] = url;
            SaveArchive(url, GetBuffer(report));
        }
        void SaveArchive(string url, byte[] buffer) {
            string tempPath = Path.ChangeExtension(StoragePath, "tmp");
            // Create a new ZIP archive.
            using (var destStream = new FileStream(tempPath, FileMode.Create))
            using (var destArchive = new ZipArchive(destStream, ZipArchiveMode.Create)) {
                // Open a ZIP archive where report files are stored.
                if (File.Exists(StoragePath)) {
                    using (var srcStream = File.OpenRead(StoragePath))
                    using (var srcArchive = new ZipArchive(srcStream, ZipArchiveMode.Read)) {
                        bool added = false;
                        // Copy all report files to a new archive.
                        // Update a file with a specified URL.
                        // If the file does not exist, create it.
                        foreach (var item in srcArchive.Entries) {
                            if (StringsEgual(item.FullName, url)) {
                                var newEntry = destArchive.CreateEntry(item.FullName);
                                using (var newStream = newEntry.Open())
                                    newStream.Write(buffer, 0, buffer.Length);
                                added = true;
                            }
                            else {
                                var newEntry = destArchive.CreateEntry(item.FullName);
                                using (var entryStream = item.Open())
                                using (var newStream = newEntry.Open())
                                    entryStream.CopyTo(newStream);
                            }
                        }
                        if (!added) {
                            var newEntry = destArchive.CreateEntry(url);
                            using (var newStream = newEntry.Open())
                                newStream.Write(buffer, 0, buffer.Length);
                        }
                    }
                }
                else {
                    var newEntry = destArchive.CreateEntry(url);
                    using (var newStream = newEntry.Open())
                        newStream.Write(buffer, 0, buffer.Length);
                }
            }
            // Replace the old ZIP archive with the new one.
            if (File.Exists(StoragePath))
                File.Delete(StoragePath);
            File.Move(tempPath, StoragePath);
        }
        byte[] GetBuffer(XtraReport report) {
            using (MemoryStream stream = new MemoryStream()) {
                report.SaveLayout(stream);
                return stream.ToArray();
            }
        }
        public override string GetNewUrl() {
            // Show the report selection dialog and return a URL for a selected report.
            StorageEditorForm form = CreateForm();
            form.textBox1.Enabled = false;
            if (form.ShowDialog() == System.Windows.Forms.DialogResult.OK)
                return form.textBox1.Text;
            return string.Empty;
        }
        StorageEditorForm CreateForm() {
            StorageEditorForm form = new StorageEditorForm();
            foreach (string item in GetUrls())
                form.listBox1.Items.Add(item);
            return form;
        }
        public override string SetNewData(XtraReport report, string defaultUrl) {
            StorageEditorForm form = CreateForm();
            form.textBox1.Text = defaultUrl;
            form.listBox1.Enabled = false;
            // Show the save dialog to get a URL for a new report.
            if (form.ShowDialog() == DialogResult.OK) {
                string url = form.textBox1.Text;
                if (!string.IsNullOrEmpty(url) && !form.listBox1.Items.Contains(url)) {
                    TypeDescriptor.GetProperties(typeof(XtraReport))["DisplayName"].SetValue(report, url);
                    SetData(report, url);
                    return url;
                }
                else {
                    MessageBox.Show("Incorrect report name", "Error",
                        MessageBoxButtons.OKCancel, MessageBoxIcon.Error);
                }
            }
            return string.Empty;
        }
        public override bool GetStandardUrlsSupported(ITypeDescriptorContext context) {
            // Always return true to confirm that the GetStandardUrls method is available.
            return true;
        }
        public override string[] GetStandardUrls(ITypeDescriptorContext context) {
            if (context != null && context.Instance is XRSubreport) {
                XRSubreport xrSubreport = context.Instance as XRSubreport;
                if (xrSubreport.RootReport !=
                    null && xrSubreport.RootReport.Extensions.TryGetValue("StorageID", out storageID)) {
                    List<string> result = GetUrlsCore(CanPassId);
                    return result.ToArray();
                }
            }
            return GetUrls();
        }
        string storageID;
        bool CanPassId(string id) {
            return id != storageID;
        }
        string[] GetUrls() {
            return GetUrlsCore(null).ToArray();
        }
        List<string> GetUrlsCore(Predicate<string> method) {
            List<string> list = new List<string>();
            if (!File.Exists(StoragePath))
                return list;
            using (var fileStream = File.OpenRead(StoragePath))
            using (var archive = new ZipArchive(fileStream, ZipArchiveMode.Read)) {
                foreach (var entry in archive.Entries)
                    if (method == null || method(entry.FullName))
                        list.Add(entry.FullName);
            }
            return list;
        }
    }
}

File: VB/ZipReportStorage.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Windows.Forms
Imports System.ComponentModel
Imports System.IO.Compression
Imports System.Collections.Generic
Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.Extensions
' ...

Namespace ReportStorageSample
    Public Class ZipReportStorage
        Inherits ReportStorageExtension

        Private Const fileName As String = "ReportStorage.zip"
        Public Sub New()
        End Sub
        Private ReadOnly Property StoragePath As String
            Get
                Dim dirName As String = Path.GetDirectoryName(Application.ExecutablePath)
                Return Path.Combine(dirName, fileName)
            End Get
        End Property
        Public Overrides Function CanSetData(url As String) As Boolean
            ' Always return true to confirm that the SetData method is available.
            Return True
        End Function
        Public Overrides Function IsValidUrl(url As String) As Boolean
            Return Not String.IsNullOrEmpty(url)
        End Function
        Public Overrides Function GetData(url As String) As Byte()
            ' Open ZIP archive.
            If Not File.Exists(StoragePath) Then
                Return New Byte() {}
            End If

            Using fileStream = File.OpenRead(StoragePath)
                Using archive = New ZipArchive(fileStream, ZipArchiveMode.Read)
                    Dim entry = archive.GetEntry(url)
                    If entry Is Nothing Then
                        Return New Byte() {}
                    End If

                    Using entryStream = entry.Open()
                        Using ms As New MemoryStream()
                            entryStream.CopyTo(ms)
                            Return ms.ToArray()
                        End Using
                    End Using
                End Using
            End Using
        End Function
        Private Shared Function StringsEgual(a As String, b As String) As Boolean
            Return String.Equals(a, b, StringComparison.OrdinalIgnoreCase)
        End Function
        Public Overrides Sub SetData(report As XtraReport, url As String)
            report.Extensions("StorageID") = url
            SaveArchive(url, GetBuffer(report))
        End Sub
        Private Sub SaveArchive(url As String, buffer As Byte())
            Dim tempPath As String = Path.ChangeExtension(StoragePath, "tmp")
            ' Create a new ZIP archive.
            Using destStream As New FileStream(tempPath, FileMode.Create)
                Using destArchive As New ZipArchive(destStream, ZipArchiveMode.Create)

                    ' Open existing archive if it exists
                    If File.Exists(StoragePath) Then
                        Using srcStream = File.OpenRead(StoragePath)
                            Using srcArchive As New ZipArchive(srcStream, ZipArchiveMode.Read)
                                Dim added As Boolean = False

                                For Each item In srcArchive.Entries
                                    Dim newEntry = destArchive.CreateEntry(item.FullName)

                                    If StringsEgual(item.FullName, url) Then
                                        Using newStream = newEntry.Open()
                                            newStream.Write(buffer, 0, buffer.Length)
                                        End Using
                                        added = True
                                    Else
                                        Using entryStream = item.Open()
                                            Using newStream = newEntry.Open()
                                                entryStream.CopyTo(newStream)
                                            End Using
                                        End Using
                                    End If
                                Next

                                If Not added Then
                                    Dim newEntry = destArchive.CreateEntry(url)
                                    Using newStream = newEntry.Open()
                                        newStream.Write(buffer, 0, buffer.Length)
                                    End Using
                                End If
                            End Using
                        End Using
                    Else
                        Dim newEntry = destArchive.CreateEntry(url)
                        Using newStream = newEntry.Open()
                            newStream.Write(buffer, 0, buffer.Length)
                        End Using
                    End If
                End Using
            End Using
            ' Replace the old ZIP archive with the new one.
            If File.Exists(StoragePath) Then
                File.Delete(StoragePath)
            End If
            File.Move(tempPath, StoragePath)
        End Sub
        Private Function GetBuffer(report As XtraReport) As Byte()
            Using stream As New MemoryStream()
                report.SaveLayout(stream)
                Return stream.ToArray()
            End Using
        End Function
        Public Overrides Function GetNewUrl() As String
            Dim form As StorageEditorForm = CreateForm()
            form.textBox1.Enabled = False

            If form.ShowDialog() = DialogResult.OK Then
                Return form.textBox1.Text
            End If
            Return String.Empty
        End Function
        Private Function CreateForm() As StorageEditorForm
            Dim form As New StorageEditorForm()
            For Each item As String In GetUrls()
                form.listBox1.Items.Add(item)
            Next

            Return form
        End Function
        Public Overrides Function SetNewData(report As XtraReport, defaultUrl As String) As String
            Dim form As StorageEditorForm = CreateForm()
            form.textBox1.Text = defaultUrl
            form.listBox1.Enabled = False

            If form.ShowDialog() = DialogResult.OK Then
                Dim url As String = form.textBox1.Text

                If Not String.IsNullOrEmpty(url) AndAlso Not form.listBox1.Items.Contains(url) Then
                    TypeDescriptor.GetProperties(GetType(XtraReport))("DisplayName").SetValue(report, url)
                    SetData(report, url)
                    Return url
                Else
                    MessageBox.Show("Incorrect report name", "Error",
                                    MessageBoxButtons.OKCancel, MessageBoxIcon.[Error])
                End If
            End If
            Return String.Empty
        End Function
        Public Overrides Function GetStandardUrlsSupported(context As ITypeDescriptorContext) As Boolean
            ' Always return true to confirm that the GetStandardUrls method is available.
            Return True
        End Function
        Public Overrides Function GetStandardUrls(context As ITypeDescriptorContext) As String()
            If context IsNot Nothing AndAlso TypeOf context.Instance Is XRSubreport Then
                Dim xrSubreport As XRSubreport = CType(context.Instance, XRSubreport)

                If xrSubreport.RootReport IsNot Nothing AndAlso
                   xrSubreport.RootReport.Extensions.TryGetValue("StorageID", storageID) Then

                    Dim result As List(Of String) = GetUrlsCore(AddressOf CanPassId)
                    Return result.ToArray()
                End If
            End If
            Return GetUrls()
        End Function
        Private storageID As String
        Private Function CanPassId(id As String) As Boolean
            Return id <> storageID
        End Function
        Private Function GetUrls() As String()
            Return GetUrlsCore(Nothing).ToArray()
        End Function
        Private Function GetUrlsCore(method As Predicate(Of String)) As List(Of String)
            Dim list As New List(Of String)()

            If Not File.Exists(StoragePath) Then
                Return list
            End If

            Using fileStream = File.OpenRead(StoragePath)
                Using archive As New ZipArchive(fileStream, ZipArchiveMode.Read)
                    For Each entry In archive.Entries
                        If method Is Nothing OrElse method(entry.FullName) Then
                            list.Add(entry.FullName)
                        End If
                    Next
                End Using
            End Using

            Return list
        End Function
    End Class
End Namespace
